apps/typegpu-docs/src/components/design/VectorSlider.tsx

```diff
@@ -0,0 +1,89 @@
+import * as RadixSlider from '@radix-ui/react-slider';
+import type * as d from 'typegpu/data';
+
+type Props<T extends d.AnyVecInstance> = {
+  min: T;
+  max: T;
+  step: T;
+  value: T;
+  onChange: (value: T) => void;
+};
+
+export function VectorSlider<T extends d.AnyVecInstance>({
+  min,
+  max,
+  step,
+  value,
+  onChange,
+}: Props<T>) {
+  const components: Array<'x' | 'y' | 'z' | 'w'> = [];
+  if ('x' in min) components.push('x');
+  if ('y' in min) components.push('y');
+  if ('z' in min) components.push('z');
+  if ('w' in min) components.push('w');
+
+  const getComponentValue = (vec: T, comp: 'x' | 'y' | 'z' | 'w'): number =>
+    comp in vec
+      ? ((vec as unknown as Record<'x' | 'y' | 'z' | 'w', number | undefined>)[
+          comp
+        ] ?? 0)
+      : 0;
+
+  const handleComponentChange = (
+    componentName: 'x' | 'y' | 'z' | 'w',
+    newValue: number,
+  ) => {
+    const newVec: Partial<d.v4f> = {};
+    if ('x' in value) newVec.x = value.x;
+    if ('y' in value) newVec.y = value.y;
+    if ('z' in value) newVec.z = value.z;
+    if ('w' in value) newVec.w = value.w;
```

Comment:
my instant idea was doing `const newVec = { ...value };`, but it doesn't really work the first time, because of how much we lie with types. imo this is hanging by a thread, but mostly because of our complicated vector implementation

Reply:
maybe a new `value.copy()` api could help here